File: src/game/quests.ts
import type { GameState } from "../types";
import { isoDay, pushMessage } from "./state";

export interface QuestDef {
  id: string;
  name: string;
  description: string;
  target: number;
  rewards: { gems?: number; gold?: number; scrolls?: number; dust?: number; orbs?: number };
}

export const DAILY_QUESTS: QuestDef[] = [
  {
    id: "campaign-3",
    name: "Campaign Commander",
    description: "Complete 3 campaign stages.",
    target: 3,
    rewards: { gold: 5000, dust: 10 },
  },
  {
    id: "pulls-1",
    name: "Fateful Summons",
    description: "Perform 1 gacha pull.",
    target: 1,
    rewards: { gems: 40, orbs: 2 },
  },
  {
    id: "arena-2",
    name: "Arena Challenger",
    description: "Fight 2 arena matches.",
    target: 2,
    rewards: { gems: 30, scrolls: 1 },
  },
  {
    id: "idle-claim",
    name: "Diligent Captain",
    description: "Claim your idle rewards.",
    target: 1,
    rewards: { gold: 2000, dust: 5 },
  },
  {
    id: "level-hero",
    name: "Trainer",
    description: "Level up any hero 5 times.",
    target: 5,
    rewards: { gold: 3000, gems: 20 },
  },
];

export const COMPLETION_REWARD = { gems: 100, scrolls: 1 };

export function ensureDailyRollover(state: GameState): GameState {
  const today = isoDay(Date.now());
  if (state.quests.day !== today) {
    return {
      ...state,
      quests: { day: today, progress: {}, claimed: {}, completionClaimed: false },
      arena: { ...state.arena, ticketsUsed: 0, ticketDay: today },
    };
  }
  return state;
}

export function bumpQuest(state: GameState, questId: string, by = 1): GameState {
  const s = ensureDailyRollover(state);
  const cur = s.quests.progress[questId] ?? 0;
  const def = DAILY_QUESTS.find((q) => q.id === questId);
  if (!def) return s;
  const next = Math.min(def.target, cur + by);
  return {
    ...s,
    quests: { ...s.quests, progress: { ...s.quests.progress, [questId]: next } },
  };
}

export function canClaimQuest(state: GameState, questId: string): boolean {
  const def = DAILY_QUESTS.find((q) => q.id === questId);
  if (!def) return false;
  if (state.quests.claimed[questId]) return false;
  const cur = state.quests.progress[questId] ?? 0;
  return cur >= def.target;
}

export function claimQuest(state: GameState, questId: string): GameState {
  if (!canClaimQuest(state, questId)) return state;
  const def = DAILY_QUESTS.find((q) => q.id === questId)!;
  const r = def.rewards;
  const s: GameState = {
    ...state,
    currency: {
      ...state.currency,
      gems: state.currency.gems + (r.gems ?? 0),
      gold: state.currency.gold + (r.gold ?? 0),
      heroicScrolls: state.currency.heroicScrolls + (r.scrolls ?? 0),
      dust: state.currency.dust + (r.dust ?? 0),
      prophetOrbs: state.currency.prophetOrbs + (r.orbs ?? 0),
    },
    quests: { ...state.quests, claimed: { ...state.quests.claimed, [questId]: true } },
  };
  return pushMessage(s, `Quest complete: ${def.name}`, "reward");
}

export function allQuestsCompleted(state: GameState): boolean {
  return DAILY_QUESTS.every((q) => (state.quests.progress[q.id] ?? 0) >= q.target);
}

export function canClaimCompletion(state: GameState): boolean {
  return allQuestsCompleted(state) && !state.quests.completionClaimed;
}

export function claimCompletion(state: GameState): GameState {
  if (!canClaimCompletion(state)) return state;
  const s: GameState = {
    ...state,
    currency: {
      ...state.currency,
      gems: state.currency.gems + COMPLETION_REWARD.gems,
      heroicScrolls: state.currency.heroicScrolls + COMPLETION_REWARD.scrolls,
    },
    quests: { ...state.quests, completionClaimed: true },
  };
  return pushMessage(s, "All daily quests complete! Bonus claimed.", "reward");
}

File: android/app/src/main/java/com/herobrawl/game/engine/Quests.kt
package com.herobrawl.game.engine

import com.herobrawl.game.model.GameState
import java.time.LocalDate
import java.time.ZoneOffset

data class QuestDef(
    val id: String,
    val name: String,
    val description: String,
    val target: Int,
    val rewardGems: Long = 0,
    val rewardGold: Long = 0,
    val rewardScrolls: Int = 0,
    val rewardDust: Long = 0,
    val rewardOrbs: Long = 0,
    val rewardFragments: Long = 0,
)

object Quests {
    val dailies: List<QuestDef> = listOf(
        QuestDef("campaign-3", "Campaign Commander", "Complete 3 campaign stages.", 3, rewardGold = 5_000, rewardDust = 10),
        QuestDef("pulls-1", "Fateful Summons", "Perform 1 gacha pull.", 1, rewardGems = 40, rewardOrbs = 2),
        QuestDef("arena-2", "Arena Challenger", "Fight 2 arena matches.", 2, rewardGems = 30, rewardScrolls = 1),
        QuestDef("idle-claim", "Diligent Captain", "Claim your idle rewards.", 1, rewardGold = 2_000, rewardDust = 5),
        QuestDef("level-hero", "Trainer", "Level up any hero 5 times.", 5, rewardGold = 3_000, rewardGems = 20),
        QuestDef("stone-equip", "Stonebinder", "Equip any stone once.", 1, rewardFragments = 30, rewardGems = 15), // new
    )

    const val completionGems = 100L
    const val completionScrolls = 1

    fun today(now: Long): String =
        LocalDate.ofInstant(java.time.Instant.ofEpochMilli(now), ZoneOffset.UTC).toString()

    fun ensureRollover(state: GameState, now: Long): GameState {
        val d = today(now)
        if (state.quests.day == d && state.arena.ticketDay == d) return state
        return state.copy(
            quests = state.quests.copy(
                day = d, progress = emptyMap(), claimed = emptyMap(), completionClaimed = false,
            ),
            arena = state.arena.copy(ticketsUsed = 0, ticketDay = d),
        )
    }

    fun bump(state: GameState, id: String, by: Int = 1): GameState {
        val def = dailies.firstOrNull { it.id == id } ?: return state
        val cur = state.quests.progress[id] ?: 0
        val next = minOf(def.target, cur + by)
        return state.copy(
            quests = state.quests.copy(progress = state.quests.progress + (id to next))
        )
    }

    fun canClaim(state: GameState, id: String): Boolean {
        val def = dailies.firstOrNull { it.id == id } ?: return false
        if (state.quests.claimed[id] == true) return false
        return (state.quests.progress[id] ?: 0) >= def.target
    }

    fun claim(state: GameState, id: String): GameState {
        if (!canClaim(state, id)) return state
        val def = dailies.first { it.id == id }
        return state.copy(
            currency = state.currency.copy(
                gems = state.currency.gems + def.rewardGems,
                gold = state.currency.gold + def.rewardGold,
                heroicScrolls = state.currency.heroicScrolls + def.rewardScrolls,
                dust = state.currency.dust + def.rewardDust,
                prophetOrbs = state.currency.prophetOrbs + def.rewardOrbs,
                stoneFragments = state.currency.stoneFragments + def.rewardFragments,
            ),
            quests = state.quests.copy(claimed = state.quests.claimed + (id to true)),
        )
    }

    fun allCompleted(state: GameState): Boolean =
        dailies.all { (state.quests.progress[it.id] ?: 0) >= it.target }

    fun canClaimCompletion(state: GameState): Boolean =
        allCompleted(state) && !state.quests.completionClaimed

    fun claimCompletion(state: GameState): GameState {
        if (!canClaimCompletion(state)) return state
        return state.copy(
            currency = state.currency.copy(
                gems = state.currency.gems + completionGems,
                heroicScrolls = state.currency.heroicScrolls + completionScrolls,
            ),
            quests = state.quests.copy(completionClaimed = true),
        )
    }
}
